File: app/src/main/java/com/example/todo4/adapter/TodoListAdapter.java
package com.example.todo4.adapter;

import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.recyclerview.widget.RecyclerView;

import com.example.todo4.R;
import com.example.todo4.model.TodoModel;

import java.util.List;

public class TodoListAdapter extends RecyclerView.Adapter<TodoListAdapter.ViewHolder> {
    private List<TodoModel> todoList;

    public TodoListAdapter(List<TodoModel> todoList) {
        this.todoList = todoList;
    }

    public static class ViewHolder extends RecyclerView.ViewHolder {
        TextView title;
        TextView detail;

        public ViewHolder(@NonNull View itemView) {
            super(itemView);
            title = itemView.findViewById(R.id.item_title);
            detail = itemView.findViewById(R.id.item_detail);
        }
    }

    @NonNull
    @Override
    public TodoListAdapter.ViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        View view = LayoutInflater.from(parent.getContext()).inflate(R.layout.todo_item, parent, false);
        return new ViewHolder(view);
    }

    @Override
    public void onBindViewHolder(@NonNull TodoListAdapter.ViewHolder holder, int position) {
        TodoModel todo = todoList.get(position);
        holder.title.setText(todo.getTitle());
        holder.detail.setText(todo.getDetail());
    }

    @Override
    public int getItemCount() {
        return todoList.size();
    }


}

File: app/src/main/java/com/example/todo4/TodoListAdapter.kt
package com.example.todo4

import android.view.LayoutInflater
import android.view.ViewGroup
import androidx.recyclerview.widget.DiffUtil
import androidx.recyclerview.widget.ListAdapter
import androidx.recyclerview.widget.RecyclerView
import com.example.todo4.data.Todo
import com.example.todo4.databinding.TodoItemBinding
import java.text.SimpleDateFormat

class TodoListAdapter(private val onItemClicked: (Todo) -> Unit) :
    ListAdapter<Todo, TodoListAdapter.ItemViewHolder>(DiffCallback) {
    override fun onCreateViewHolder(
        parent: ViewGroup,
        viewType: Int
    ): ItemViewHolder {
        return ItemViewHolder(
            TodoItemBinding.inflate(LayoutInflater.from(parent.context))
        )
    }

    override fun onBindViewHolder(holder: ItemViewHolder, position: Int) {
        val current = getItem(position)
        holder.itemView.setOnClickListener {
            onItemClicked(current)
        }
        holder.bind(current)
    }

    class ItemViewHolder(private var binding: TodoItemBinding) :
        RecyclerView.ViewHolder(binding.root) {
        fun bind(todo: Todo) {
            binding.itemTitle.text = todo.title
            binding.itemDetail.text = todo.detail
            if (todo.dueDate != null) {
                val dateFormat = SimpleDateFormat("yyyy-MM-dd")

                binding.listTvDeu.text = dateFormat.format(todo.dueDate)
            }
        }

    }

    companion object {
        private val DiffCallback = object : DiffUtil.ItemCallback<Todo>() {
            override fun areItemsTheSame(oldTodo: Todo, newTodo: Todo): Boolean {
                return oldTodo === newTodo
            }

            override fun areContentsTheSame(oldTodo: Todo, newTodo: Todo): Boolean {
                return oldTodo.id == newTodo.id
            }
        }
    }
}